File: trunk/simulations-java/simulations/moving-man/scala-src/edu/colorado/phet/movingman/ladybug/controlpanel/ModePanel.scala
package edu.colorado.phet.movingman.ladybug.controlpanel

import _root_.scala.swing.Panel
import javax.swing.BoxLayout._
import javax.swing.{BoxLayout, JPanel, JComponent}
import model.LadybugModel
import edu.colorado.phet.movingman.ladybug.swing.MyRadioButton
import java.awt.Color._

class ModePanel(model: LadybugModel) extends JPanel {
    setLayout(new BoxLayout(this, Y_AXIS))
    def color(b: Boolean) = if (b) red else black

    val recordingButton = addNewControl{new MyRadioButton("Recording", model.setRecord(true), {model.isRecord}, model.addListener)}
    val playbackButton = addNewControl{new MyRadioButton("Playback", model.setRecord(false), {model.isPlayback}, model.addListener)}

    def addAndInvoke(addListener: (() => Unit) => Unit)(updateFunction: () => Unit) = {
        addListener(updateFunction)
        updateFunction
    }

    addAndInvoke(model.addListener){
        () => {
            recordingButton.peer.setForeground(color(recordingButton.peer.isSelected && !model.isPaused))
            playbackButton.peer.setForeground(color(playbackButton.peer.isSelected && !model.isPaused))
        }
    }

    //a control structure that (1) creates a swing component and (2) automatically adds it
    def addNewControl(m: => MyRadioButton): MyRadioButton = {
        val component = m
        add(component.peer)
        component
    }
}

File: branches/test-branch-44224/simulations-java/common/record-and-playback/src/edu/colorado/phet/recordandplayback/gui/ModePanel.java
package edu.colorado.phet.recordandplayback.gui;

import edu.colorado.phet.common.phetcommon.resources.PhetCommonResources;
import edu.colorado.phet.common.phetcommon.util.SimpleObserver;
import edu.colorado.phet.common.phetcommon.view.util.PhetFont;
import edu.colorado.phet.recordandplayback.model.RecordAndPlaybackModel;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

/**
 * The ModePanel indicates whether the sim is in record or playback mode; the record or playback text is highlighted
 * in red if the sim is indeed recording or playing back (may not be if recording time is full or sim is paused).
 *
 * @author Sam Reid
 * @param <T> the model type
 */
public class ModePanel<T> extends JPanel {

    public ModePanel(final RecordAndPlaybackModel<T> model) {
        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
        setBackground(new Color(0, 0, 0, 0));
        final JRadioButton recordingButton = new JRadioButton(PhetCommonResources.getString("Common.record"), model.isRecord());
        recordingButton.setBackground(new Color(0, 0, 0, 0));
        recordingButton.setFont(new PhetFont(15, true));
        recordingButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                model.setRecord(true);
            }
        });

        final JRadioButton playbackButton = new JRadioButton(PhetCommonResources.getString("Common.playback"), model.isPlayback());
        playbackButton.setFont(new PhetFont(15, true));
        playbackButton.setBackground(new Color(0, 0, 0, 0));
        playbackButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                model.setRecord(false);
                model.rewind();
                model.setPaused(true);
            }
        });

        add(recordingButton);
        add(playbackButton);

        model.addObserver(new SimpleObserver() {
            public void update() {
                recordingButton.setForeground(color(recordingButton.isSelected() && !model.isPaused() && !model.isRecordingFull()));
                playbackButton.setForeground(color(playbackButton.isSelected() && !model.isPaused()));
                recordingButton.setSelected(model.isRecord());
                playbackButton.setSelected(model.isPlayback());
            }
        });
    }

    private Color color(boolean b) {
        if (b) return Color.red;
        else return Color.black;
    }
}